File: backend/app/models/document.py
import uuid
import enum
from datetime import datetime
import uuid as _uuid
from sqlalchemy import Column, String, DateTime, Text, ForeignKey, Enum as SQLEnum, BigInteger, Integer, Boolean, Float, Date, Index, JSON
from sqlalchemy.orm import relationship
from app.db.session import Base


class _UUIDType:
    """Cross-dialect UUID type - stores as String(36) internally, works with SQLite and PostgreSQL"""
    pass


def _uuid_column(**kwargs):
    return Column(String(36), default=lambda: str(_uuid.uuid4()), **kwargs)


def _json_column(**kwargs):
    return Column(JSON, **kwargs)


class DocumentType(str, enum.Enum):
    PDF = "PDF"
    EXCEL = "Excel"
    IMAGE = "Image"
    WORD = "Word"
    OTHER = "Other"


class DocumentCategory(str, enum.Enum):
    SALES_TAX_RETURN = "Sales Tax Return"
    SECTION_236H = "236H"
    SECTION_153 = "153"
    SECTION_165 = "165"
    KPRA = "KPRA"
    INCOME_TAX_RETURN = "Income Tax Return"
    WORKING_FILE = "Working File"
    NOTICE = "Notice"
    OTHER = "Other"


class FilingStatus(str, enum.Enum):
    FILED = "Filed"
    PENDING = "Pending"
    NOT_FILED = "Not Filed"
    OVERDUE = "Overdue"
    MISSING = "Missing"
    UPLOADED = "Uploaded"


class DocumentActivityType(str, enum.Enum):
    VIEW = "view"
    DOWNLOAD = "download"
    PRINT = "print"
    PREVIEW = "preview"
    UPLOAD = "upload"
    DELETE = "delete"
    RENAME = "rename"
    MOVE = "move"
    COPY = "copy"
    RESTORE = "restore"
    SHARE = "share"


class Document(Base):
    __tablename__ = "documents"

    # Primary Key
    id = Column(String(36), primary_key=True, default=lambda: str(_uuid.uuid4()))

    # Foreign Key
    client_id = Column(String(36), ForeignKey("clients.id", ondelete="CASCADE"), nullable=False)

    # File Information
    file_name = Column(Text, nullable=False)
    original_file_name = Column(Text, nullable=False)
    file_extension = Column(String(20), nullable=False)
    file_size = Column(BigInteger, nullable=False)
    file_path = Column(Text, nullable=False)
    file_type = Column(SQLEnum(DocumentType), nullable=False)

    # Classification
    doc_category = Column(SQLEnum(DocumentCategory), nullable=True)
    classification_method = Column(String(50), nullable=True)  # pattern, content, client_context, manual, fallback
    classification_confidence = Column(Float, default=0.0)

    # Tax Period
    tax_year = Column(Integer, nullable=True)
    tax_month = Column(Integer, nullable=True)  # 1-12

    # Compliance
    filing_status = Column(SQLEnum(FilingStatus), nullable=True)
    is_missing = Column(Boolean, default=False)

    # Dates
    document_date = Column(Date, nullable=True)  # Document issuance date
    expiry_date = Column(Date, nullable=True)  # For notices/returns with deadlines
    upload_date = Column(DateTime, default=datetime.utcnow, nullable=False)

    # Audit
    uploaded_by = Column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    deleted_at = Column(DateTime, nullable=True)
    is_deleted = Column(Boolean, default=False)

    # Version Control
    version = Column(Integer, default=1)
    parent_document_id = Column(String(36), ForeignKey("documents.id", ondelete="SET NULL"), nullable=True)

    # Metadata
    notes = Column(Text, nullable=True)
    tags = Column(JSON, nullable=True)
    custom_metadata = Column(JSON, default={})

    # Batch Tracking
    batch_id = Column(String(36), nullable=True)
    checksum = Column(String(64), nullable=True)  # SHA-256

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    client = relationship("Client", backref="documents")
    uploader = relationship("User", backref="uploaded_documents", foreign_keys=[uploaded_by])
    parent_document = relationship("Document", remote_side=[id], backref="child_documents")

    def __repr__(self):
        return f"<Document(id={self.id}, file_name='{self.file_name}', category='{self.doc_category}')>"


class DocumentActivityLog(Base):
    __tablename__ = "document_activity_log"

    id = Column(String(36), primary_key=True, default=lambda: str(_uuid.uuid4()))
    document_id = Column(String(36), ForeignKey("documents.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    activity_type = Column(SQLEnum(DocumentActivityType), nullable=False)
    ip_address = Column(String(45), nullable=True)
    user_agent = Column(Text, nullable=True)
    activity_metadata = Column("metadata", JSON, default={})
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    # Relationships
    document = relationship("Document", backref="activity_logs")
    user = relationship("User", backref="document_activities")

    def __repr__(self):
        return f"<DocumentActivityLog(document_id={self.document_id}, type='{self.activity_type}')>"


class SavedFilter(Base):
    __tablename__ = "saved_filters"

    id = Column(String(36), primary_key=True, default=lambda: str(_uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(200), nullable=False)
    filter_config = Column(JSON, nullable=False)
    is_shared = Column(Boolean, default=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    user = relationship("User", backref="saved_filters")

    def __repr__(self):
        return f"<SavedFilter(name='{self.name}', user_id={self.user_id})>"

File: frontend/src/features/documents/types/document.ts
// ============================================================================
// Document Module Types
// ============================================================================

// --- Enums ---

export type DocumentType = 'PDF' | 'Excel' | 'Image' | 'Word' | 'Other';

export type DocumentCategory =
  | 'Sales Tax Return'
  | '236H'
  | '153'
  | '165'
  | 'KPRA'
  | 'Income Tax Return'
  | 'Working File'
  | 'Notice'
  | 'Other';

export type FilingStatus =
  | 'Filed'
  | 'Pending'
  | 'Not Filed'
  | 'Overdue'
  | 'Missing'
  | 'Uploaded';

export type DocumentActivityType =
  | 'view'
  | 'download'
  | 'print'
  | 'preview'
  | 'upload'
  | 'delete'
  | 'rename'
  | 'move'
  | 'copy'
  | 'restore'
  | 'share';

export type ViewMode = 'grid' | 'list';

export type SortField = 'file_name' | 'upload_date' | 'file_size' | 'client_name' | 'doc_category' | 'filing_status';
export type SortOrder = 'asc' | 'desc';

// --- Core Types ---

export interface Document {
  id: string;
  client_id: string;
  file_name: string;
  original_file_name: string;
  file_extension: string;
  file_size: number;
  file_path: string;
  file_type: DocumentType;

  // Classification
  doc_category: DocumentCategory | null;
  classification_method: string | null;
  classification_confidence: number;

  // Tax Period
  tax_year: number | null;
  tax_month: number | null;

  // Compliance
  filing_status: FilingStatus | null;
  is_missing: boolean;

  // Dates
  document_date: string | null;
  expiry_date: string | null;
  upload_date: string;

  // Audit
  uploaded_by: string | null;
  deleted_at: string | null;
  is_deleted: boolean;

  // Version
  version: number;
  parent_document_id: string | null;

  // Metadata
  notes: string | null;
  tags: string[] | null;
  custom_metadata: Record<string, unknown>;

  // Batch
  batch_id: string | null;
  checksum: string | null;

  // Timestamps
  created_at: string;
  updated_at: string;

  // Computed (from joins)
  client_name?: string;
  client_ntn?: string;
  client_cnic?: string;
  client_strn?: string;
  uploader_name?: string;
}

// --- API Responses ---

export interface DocumentListResponse {
  success: boolean;
  data: Document[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface DocumentStats {
  total_documents: number;
  total_pdf: number;
  total_excel: number;
  recent_uploads_24h: number;
  recent_uploads_7d: number;
  missing_documents: number;
  uploads_this_month: number;
  uploads_previous_month: number;
  monthly_trend: MonthlyTrend[];
  total_clients_with_gaps: number;
}

export interface MonthlyTrend {
  month: string;
  count: number;
}

export interface DocumentStatsResponse {
  success: boolean;
  data: DocumentStats;
}

// --- Folder Types ---

export interface FolderNode {
  name: string;
  path: string;
  type: 'folder' | 'file';
  document_count: number;
  children: FolderNode[];
  is_expanded?: boolean;
}

export interface FolderContentsResponse {
  success: boolean;
  data: Document[];
  folder_path: string;
  total: number;
  page: number;
  limit: number;
}

// --- Compliance Types ---

export interface ComplianceDocumentStatus {
  category: DocumentCategory;
  status: 'uploaded' | 'missing' | 'pending' | 'not_required';
  document_id: string | null;
  filing_status: FilingStatus | null;
}

export interface ComplianceMonthStatus {
  month: number;
  month_name: string;
  documents: ComplianceDocumentStatus[];
}

export interface ComplianceSummary {
  total_required: number;
  uploaded: number;
  missing: number;
  pending: number;
  compliance_percentage: number;
}

export interface ComplianceStatusResponse {
  success: boolean;
  data: {
    client: { id: string; name: string; ntn: string | null };
    year: number;
    compliance: ComplianceMonthStatus[];
    summary: ComplianceSummary;
  };
}

export interface MissingDocument {
  client_id: string
  client_name: string
  client_ntn: string | null
  required_type: DocumentCategory
  tax_year: number
  tax_month: number
  deadline: string
  days_overdue: number
  is_overdue: boolean
}

export interface MissingDocumentsResponse {
  success: boolean;
  data: MissingDocument[];
  total: number;
}

export interface MissingCountResponse {
  success: boolean;
  data: {
    total_missing: number;
    overdue: number;
    upcoming: number;
  };
}

// --- Activity Types ---

export interface DocumentActivity {
  id: string;
  document_id: string;
  user_id: string | null;
  activity_type: DocumentActivityType;
  ip_address: string | null;
  user_agent: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  user_name?: string;
}

// --- Filter Types ---

export interface DocumentFilters {
  // Search
  q?: string;

  // Category filters
  doc_category?: DocumentCategory[];
  tax_year?: number;
  tax_month?: number[];
  client_id?: string[];
  filing_status?: FilingStatus | null;
  file_type?: DocumentType | null;

  // Date range
  upload_date_from?: string;
  upload_date_to?: string;

  // File size
  file_size_min?: number;
  file_size_max?: number;

  // Compliance
  is_missing?: boolean;

  // Sorting
  sort_by?: SortField;
  sort_order?: SortOrder;

  // Pagination
  page?: number;
  limit?: number;

  // Folder
  folder_path?: string;
}

export interface SavedFilter {
  id: string;
  user_id: string;
  name: string;
  filter_config: DocumentFilters;
  is_shared: boolean;
  created_at: string;
  updated_at: string;
}

// --- Upload Types ---

export interface UploadProgress {
  file: File;
  progress: number;
  status: 'pending' | 'uploading' | 'success' | 'error';
  result?: Document;
  error?: string;
}

export interface BulkUploadRequest {
  client_id: string;
  doc_category?: DocumentCategory;
  tax_year?: number;
  tax_month?: number;
  overwrite?: boolean;
}

export interface BulkUploadResult {
  success: Document[];
  errors: { file_name: string; error: string }[];
  skipped: { file_name: string; reason: string }[];
}

// --- Action Types ---

export interface RenameRequest {
  file_name: string;
}

export interface MoveRequest {
  client_id: string;
  folder_path: string;
}

export interface CopyRequest {
  client_id: string;
  folder_path: string;
}

// --- Document Type Utilities ---

export const DOCUMENT_CATEGORY_OPTIONS: { value: DocumentCategory; label: string; color: string }[] = [
  { value: 'Sales Tax Return', label: 'Sales Tax Return', color: 'bg-blue-100 text-blue-700' },
  { value: '236H', label: '236H', color: 'bg-purple-100 text-purple-700' },
  { value: '153', label: '153', color: 'bg-indigo-100 text-indigo-700' },
  { value: '165', label: '165', color: 'bg-rose-100 text-rose-700' },
  { value: 'KPRA', label: 'KPRA', color: 'bg-cyan-100 text-cyan-700' },
  { value: 'Income Tax Return', label: 'Income Tax Return', color: 'bg-emerald-100 text-emerald-700' },
  { value: 'Working File', label: 'Working File', color: 'bg-slate-100 text-slate-700' },
  { value: 'Notice', label: 'Notice', color: 'bg-amber-100 text-amber-700' },
  { value: 'Other', label: 'Other', color: 'bg-gray-100 text-gray-700' },
];

export const FILING_STATUS_OPTIONS: { value: FilingStatus; label: string; color: string }[] = [
  { value: 'Filed', label: 'Filed', color: 'bg-emerald-100 text-emerald-700' },
  { value: 'Uploaded', label: 'Uploaded', color: 'bg-emerald-100 text-emerald-700' },
  { value: 'Pending', label: 'Pending', color: 'bg-amber-100 text-amber-700' },
  { value: 'Not Filed', label: 'Not Filed', color: 'bg-slate-100 text-slate-700' },
  { value: 'Overdue', label: 'Overdue', color: 'bg-red-100 text-red-700' },
  { value: 'Missing', label: 'Missing', color: 'bg-red-100 text-red-700' },
];

export const MONTHS = [
  { value: 1, label: 'January' },
  { value: 2, label: 'February' },
  { value: 3, label: 'March' },
  { value: 4, label: 'April' },
  { value: 5, label: 'May' },
  { value: 6, label: 'June' },
  { value: 7, label: 'July' },
  { value: 8, label: 'August' },
  { value: 9, label: 'September' },
  { value: 10, label: 'October' },
  { value: 11, label: 'November' },
  { value: 12, label: 'December' },
];

// Helper functions
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + ' ' + sizes[i];
}

export function getFileTypeIcon(fileType: DocumentType): string {
  switch (fileType) {
    case 'PDF': return 'FileText';
    case 'Excel': return 'Sheet';
    case 'Image': return 'Image';
    case 'Word': return 'FileText';
    default: return 'File';
  }
}

export function getFileTypeColor(fileType: DocumentType): string {
  switch (fileType) {
    case 'PDF': return 'text-red-500';
    case 'Excel': return 'text-emerald-500';
    case 'Image': return 'text-blue-500';
    case 'Word': return 'text-blue-600';
    default: return 'text-slate-500';
  }
}

export function getFileTypeBgColor(fileType: DocumentType): string {
  switch (fileType) {
    case 'PDF': return 'bg-red-50';
    case 'Excel': return 'bg-emerald-50';
    case 'Image': return 'bg-blue-50';
    case 'Word': return 'bg-blue-50';
    default: return 'bg-slate-50';
  }
}